filebeat/input/v2/statestore/resource.go

```diff
@@ -0,0 +1,313 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package statestore
+
+import (
+	"errors"
+	"runtime"
+
+	"github.com/elastic/beats/libbeat/common/transform/typeconv"
+	"github.com/elastic/beats/libbeat/registry"
+)
+
+// Resource is used to lock and modify a resource its registry contents in a
+// store.
+type Resource struct {
+	store    *Store
+	isLocked bool
+
+	key   ResourceKey
+	entry *resourceEntry
+}
+
+func newResource(store *Store, key ResourceKey) *Resource {
+	res := &Resource{
+		store:    store,
+		isLocked: false,
+		key:      key,
+	}
+
+	// in case we miss an unlock operation (programmer error or panic that hash
+	// been caught) we set a finalizer to eventually free the resource.
+	// The Unlock operation will unsert the finalizer.
+	runtime.SetFinalizer(res, (*Resource).finalize)
+	return res
+}
+
+func (r *Resource) close() {
+	runtime.SetFinalizer(r, nil)
+	r.finalize()
+}
+
+func (r *Resource) finalize() {
+	defer r.unlink()
+
+	if r.IsLocked() {
+		r.doUnlock()
+	}
+}
+
+func (r *Resource) link(create bool) {
+	if r.entry != nil {
+		return
+	}
+
+	store := r.store
+	store.resourcesMux.Lock()
+	defer store.resourcesMux.Unlock()
+
+	entry := store.find(r.key)
+	if entry == nil && create {
+		entry = store.create(r.key)
+	}
+	r.entry = entry
+}
+
+// unlink removes the pointer to the memory backed resource.
+// If the in memory resource entry is not used by any other Resource,
+// then we will remove it from the table.
+func (r *Resource) unlink() {
+	if r.entry == nil {
+		return
+	}
+
+	entry := r.entry
+	r.entry = nil
+
+	store := r.store
+	store.resourcesMux.Lock()
+	defer store.resourcesMux.Unlock()
+	if entry.Release() {
+		store.remove(r.key)
+	}
+}
+
+// Lock locks a resource held by the store. It blocks until the lock becomes
+// available.
+func (r *Resource) Lock() {
+	checkNotLocked(r.IsLocked())
+
+	r.link(true)
+	r.entry.Lock()
+	r.isLocked = true
+}
+
+// TryLock attempts to lock the resource. It returns true if the lock has been
+// acquired successfully.
+func (r *Resource) TryLock() bool {
+	checkNotLocked(r.IsLocked())
+
+	r.link(true)
+	r.isLocked = r.entry.TryLock()
+	if !r.isLocked {
+		r.unlink()
+	}
+	return r.isLocked
+}
```

Comment:
I was thinking maybe a Lock that accept a deadline, but we can do the same logic in the consumer with the tryLock().

Reply:
TryLock is supposed to return immediately. I don't want to change Lock/Unlock as is, but some `LockWithContext` would indeed be nice. E.g. we could unblock the call on shutdown.

Idea of TryLock is similar to current 'Finished' flag in filebeat. Assumed code pattern:

```
res, ok := store.TryLock(path)
if !ok {
  return fmt.Errorf("file %v is collected by another input", path)
}
```

Reply:
Sound good, The current API allow for that flexibility so lets keep it simple and add if needed or event make function that wrap that behavior.